File: src/prompt/build_segments.zig
//! Turns detected context + git info into the ordered Segment list the
//! renderer draws. This is the adaptive core: a segment appears only when the
//! context calls for it.

const std = @import("std");
const seg = @import("segments.zig");
const ctx = @import("context.zig");
const git = @import("git.zig");

pub const Inputs = struct {
    cwd_base: []const u8, // basename of the working directory
    context: ctx.Context,
    git_info: ?git.Info,
    exit_code: u8,
    /// scratch buffer the assembled segment texts borrow from
    scratch: []u8,
};

fn langText(l: ctx.Lang) ?[]const u8 {
    return switch (l) {
        .none => null,
        .zig => "zig",
        .node => "node",
        .python => "python",
        .rust => "rust",
        .go => "go",
    };
}

/// Build the active segment list. Texts are slices into `in.scratch`.
pub fn assemble(in: Inputs) seg.List {
    var list = seg.List{};
    var off: usize = 0;

    // cwd — always.
    list.add(.{ .icon = .repo, .text = in.cwd_base });

    // git — when in a repo.
    if (in.git_info) |g| {
        const dirty_suffix = if (g.dirty > 0) blk: {
            if (std.fmt.bufPrint(in.scratch[off..], "{s} \u{25cf}{d}", .{ g.branch, g.dirty })) |s| {
                off += s.len;
                break :blk s;
            } else |_| {
                break :blk g.branch;
            }
        } else g.branch;
        list.add(.{
            .icon = .branch,
            .text = dirty_suffix,
            .state = if (g.dirty > 0) .warn else .normal,
        });
    }

    // toolchain — when a language is detected.
    if (langText(in.context.lang)) |lt| {
        list.add(.{ .icon = .toolchain, .text = lt });
    }

    // container / cluster — when present.
    if (in.context.has_container) list.add(.{ .icon = .container, .text = "docker" });
    if (in.context.has_k8s) list.add(.{ .icon = .cluster, .text = "k8s" });

    // failure — only on a non-zero exit.
    if (in.exit_code != 0) {
        const exit_text = if (std.fmt.bufPrint(in.scratch[off..], "{d}", .{in.exit_code})) |s| blk: {
            off += s.len;
            break :blk s;
        } else |_| "?";
        list.add(.{ .icon = .err, .text = exit_text, .state = .err });
    }

    return list;
}

const testing = std.testing;
const icons = @import("icons.zig");

test "assemble: clean repo shows cwd + branch only" {
    var scratch: [256]u8 = undefined;
    const list = assemble(.{
        .cwd_base = "caldera-console",
        .context = .{ .in_git = true },
        .git_info = .{ .branch = "main" },
        .exit_code = 0,
        .scratch = &scratch,
    });
    try testing.expectEqual(@as(usize, 2), list.slice().len);
    try testing.expectEqual(seg.Segment{ .icon = .repo, .text = "caldera-console" }, list.slice()[0]);
}

test "assemble: dirty repo marks the git segment warn" {
    var scratch: [256]u8 = undefined;
    const list = assemble(.{
        .cwd_base = "x",
        .context = .{ .in_git = true },
        .git_info = .{ .branch = "main", .dirty = 3 },
        .exit_code = 0,
        .scratch = &scratch,
    });
    try testing.expectEqual(seg.State.warn, list.slice()[1].state);
    try testing.expect(std.mem.indexOf(u8, list.slice()[1].text, "3") != null);
}

test "assemble: a node+docker dir surfaces toolchain and container" {
    var scratch: [256]u8 = undefined;
    const list = assemble(.{
        .cwd_base = "app",
        .context = .{ .lang = .node, .has_container = true },
        .git_info = null,
        .exit_code = 0,
        .scratch = &scratch,
    });
    var saw_tool = false;
    var saw_dk = false;
    for (list.slice()) |s| {
        if (s.icon == .toolchain) saw_tool = true;
        if (s.icon == .container) saw_dk = true;
    }
    try testing.expect(saw_tool and saw_dk);
}

test "assemble: a non-zero exit adds an err segment" {
    var scratch: [256]u8 = undefined;
    const list = assemble(.{
        .cwd_base = "x",
        .context = .{},
        .git_info = null,
        .exit_code = 127,
        .scratch = &scratch,
    });
    const last = list.slice()[list.slice().len - 1];
    try testing.expectEqual(icons.Icon.err, last.icon);
    try testing.expectEqualStrings("127", last.text);
}

File: crates/anvil-prompt-core/src/build_segments.rs
//! Turns detected context + git info into the ordered Segment list the
//! renderer draws. This is the adaptive core: a segment appears only when the
//! context calls for it.

use crate::context::{Context, Lang};
use crate::git::Info as GitInfo;
use crate::icons::Icon;
use crate::segments::{List, Segment, State};

pub struct Inputs<'a> {
    /// Basename of the working directory.
    pub cwd_base: &'a str,
    pub context: Context,
    pub git_info: Option<GitInfo>,
    pub exit_code: u8,
}

fn lang_text(l: Lang) -> Option<&'static str> {
    match l {
        Lang::None => None,
        Lang::Zig => Some("zig"),
        Lang::Node => Some("node"),
        Lang::Python => Some("python"),
        Lang::Rust => Some("rust"),
        Lang::Go => Some("go"),
    }
}

/// Build the active segment list.
pub fn assemble(input: Inputs<'_>) -> List {
    let mut list = List::new();

    // cwd — always.
    list.add(Segment::new(Icon::Repo, input.cwd_base));

    // git — when in a repo.
    if let Some(g) = input.git_info {
        let (text, state) = if g.dirty > 0 {
            (format!("{} {}", g.branch, g.dirty), State::Warn)
        } else {
            (g.branch.clone(), State::Normal)
        };
        list.add(Segment::with_state(Icon::Branch, text, state));
    }

    // toolchain — when a language is detected.
    if let Some(lt) = lang_text(input.context.lang) {
        list.add(Segment::new(Icon::Toolchain, lt));
    }

    // container / cluster — when present.
    if input.context.has_container {
        list.add(Segment::new(Icon::Container, "docker"));
    }
    if input.context.has_k8s {
        list.add(Segment::new(Icon::Cluster, "k8s"));
    }

    // failure — only on a non-zero exit.
    if input.exit_code != 0 {
        let exit_text = format!("{}", input.exit_code);
        list.add(Segment::with_state(Icon::Err, exit_text, State::Err));
    }

    list
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::context::Lang;

    #[test]
    fn assemble_clean_repo_shows_cwd_and_branch_only() {
        let list = assemble(Inputs {
            cwd_base: "anvil",
            context: Context {
                in_git: true,
                ..Context::default()
            },
            git_info: Some(GitInfo {
                branch: "main".into(),
                dirty: 0,
                ahead: 0,
                behind: 0,
            }),
            exit_code: 0,
        });
        assert_eq!(list.slice().len(), 2);
        assert_eq!(list.slice()[0].icon, Icon::Repo);
        assert_eq!(list.slice()[0].text, "anvil");
    }

    #[test]
    fn assemble_dirty_repo_marks_git_segment_warn() {
        let list = assemble(Inputs {
            cwd_base: "x",
            context: Context {
                in_git: true,
                ..Context::default()
            },
            git_info: Some(GitInfo {
                branch: "main".into(),
                dirty: 3,
                ahead: 0,
                behind: 0,
            }),
            exit_code: 0,
        });
        assert_eq!(list.slice()[1].state, State::Warn);
        assert!(list.slice()[1].text.contains('3'));
    }

    #[test]
    fn assemble_node_docker_dir_surfaces_toolchain_and_container() {
        let list = assemble(Inputs {
            cwd_base: "app",
            context: Context {
                lang: Lang::Node,
                has_container: true,
                ..Context::default()
            },
            git_info: None,
            exit_code: 0,
        });
        let mut saw_tool = false;
        let mut saw_dk = false;
        for s in list.slice() {
            if s.icon == Icon::Toolchain {
                saw_tool = true;
            }
            if s.icon == Icon::Container {
                saw_dk = true;
            }
        }
        assert!(saw_tool && saw_dk);
    }

    #[test]
    fn assemble_non_zero_exit_adds_err_segment() {
        let list = assemble(Inputs {
            cwd_base: "x",
            context: Context::default(),
            git_info: None,
            exit_code: 127,
        });
        let last = list.slice().last().unwrap();
        assert_eq!(last.icon, Icon::Err);
        assert_eq!(last.text, "127");
    }
}
